luaui/Widgets/unit_factory_quota.lua

```diff
@@ -0,0 +1,253 @@
+    local best
+    local bestKey
+    for k, v in pairs(table) do
+        if not best then
+            best = v
+            bestKey = k
+        elseif f(v, best, k, bestKey) then
+            best = v
+            bestKey = k
+        end
+    end
+    return best, bestKey
+end
+
+---length of all tables
+---@param table table
+---@return integer
+function table.length(table)
+    local x = 0
+    for _ in pairs(table or {}) do
+        x = x + 1
+    end
+    return x
+end
+
+local function isFactoryUsable(factoryID)
+    local commandq = GetFactoryCommands(factoryID, 2)
+    if not commandq then
+        return true
+    end
+    return commandq and( #commandq == 0 or not (commandq[1].options.alt or (commandq[2] and commandq[2].options.alt)))
+end
+
+local function orderDequeue(unitID, buildDefID, count)
+	while count > 0 do
+		count = count - 100
+
+		GiveOrderToUnit(unitID, -buildDefID, {}, { "right", "ctrl", "shift" })
+	end
+end
+
+local function clearFactQ(factID)
+    
+	local queue = GetRealBuildQueue(unitID)
+
+	if queue ~= nil then
+		for _, buildPair in ipairs(queue) do
+			local buildUnitDefID, count = next(buildPair, nil)
+
+			orderDequeue(unitID, buildUnitDefID, count)
+		end
+	end
+end
+
+local function putInFrontOfFactQ(...)
+    local factID, cmdID, params, opts = ...
+    local commandq = GetFactoryCommands(factID, -1)
+    local altCmds = {}
+    local others = {}
+    for _, cmd in ipairs(commandq) do
+        if cmd.options.alt then
+            altCmds[#altCmds+1] = cmd
+        else
+            others[#others+1] = cmd
+        end
+    end
+
+    clearFactQ(factID)
+
+    for i = #altCmds, 1, -1 do -- do alt queue backwards
+        local cmd = altCmds[i]
+        GiveOrderToUnit(factID, cmd.id, cmd.params, cmd.options)
+    end
+    GiveOrderToUnit(...)
+    for _, cmd in ipairs(others) do
+        GiveOrderToUnit(factID, cmd.id, cmd.params, cmd.options)
+    end
+end
+
+local function insertToFactQ(...)
+    local factID, cmdID, params, opts = ...
+    local currCmd, targetID = Spring.GetUnitWorkerTask(factID)
+    local insertnormally = true
+    if targetID and Spring.GetUnitStates(factID)["repeat"] then
+        local _, _, _, _, buildProgress = Spring.GetUnitHealth(targetID)
+        if buildProgress < maxBuildProg and metalcosts[-currCmd] and (buildProgress * metalcosts[-currCmd]) < maxMetal then -- 7.5 % is the most that it is willing to cancel, and maximally 500 metal
+            insertnormally = false
+        end
+    end
+    if insertnormally then
+        GiveOrderToUnit(...)
+    else
+        putInFrontOfFactQ(...)
+    end
+end
+
+local function fillQuotas()
+    for factID, quota in pairs(quotas) do
+        if isFactoryUsable(factID) then
+            for udefid, num in pairs(quota) do
+                if num == 0 then
+                    quota[udefid] = nil
+                end
+            end
+            if table.length(quota)>0 then
+                local function isBetter(q1, q2, k1, k2)
```

Comment:
What does `isBetter` mean? What does this function do? What do the parameters represent? Please give them descriptive names.